### PR TITLE
Add GaussianSmooth as antialiasing filter in Resize

### LINKED ISSUE #4020: Resize transform produces aliasing artifacts
**Describe the bug**
The Resize transform produces aliasing artifacts. It uses the `F.interpolate` function from PyTorch, which has an antialiasing option, but that does not support 3D downsampling of volumes (5D tensors). The Resize transforms does not use the antialiasing option at all: https://github.com/Project-MONAI/MONAI/blob/a676e3876903e5799eb3d24245872ea4f3a5b152/monai/transforms/spatial/array.py#L676

This line in PyTorch source code confirms that the antialiasing option is not available for 3D volumes (trilinear interpolation):
https://github.com/pytorch/pytorch/blob/8ed6cb42ba2635319b67417b9ee2f1a0142d2c25/torch/nn/functional.py#L3851

The issue of ignoring aliasing in deep learning research is covered in this paper: https://arxiv.org/abs/2104.11222

**To Reproduce**
1. Resize an image or volume down to a small size, e.g. using a factor of 8
2. Observe the aliasing artifacts in the resized tensor

**Expected behavior**
The downsampled image should look like a smaller version of the original (no aliasing).

### PR DESCRIPTION
**Pull request by Can-Zhao**
…Resize Transform

Fixes #4020 .

### Description
Add GaussianSmooth as anti-aliasing filter in Resize Transform

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [X] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/transforms/spatial/array.py [hunk 1]
```diff
@@ -597,6 +598,70 @@ def __call__(self, img: NdarrayOrTensor) -> NdarrayOrTensor:
         return torch.flip(img, map_spatial_axes(img.ndim, self.spatial_axis))
 
 
+class GaussianSmoothing(Transform):
```
Comment (wyli) [inline monai/transforms/spatial/array.py:601]: Hi @Can-Zhao, there is a gaussian smoothing method here https://github.com/Project-MONAI/MONAI/blob/6602303578cb7f331fce7d3bb4334396e0bfee2d/monai/transforms/intensity/array.py#L1130-L1140 (without the batch dim)

and this one (trainable): https://github.com/Project-MONAI/MONAI/blob/6602303578cb7f331fce7d3bb4334396e0bfee2d/monai/networks/layers/simplelayers.py#L442

could you please reuse any of these APIs? Let me know if they are bugs/limitations with the existing ones...

File: monai/transforms/spatial/array.py [hunk 2]
```diff
@@ -618,6 +683,8 @@ class Resize(Transform):
         align_corners: This only has an effect when mode is
             'linear', 'bilinear', 'bicubic' or 'trilinear'. Default: None.
             See also: https://pytorch.org/docs/stable/generated/torch.nn.functional.interpolate.html
+        anti_aliasing: bool. Whether to apply a Gaussian filter to smooth the image prior to downsampling.
+            Only valid when downsamling. See also skimage.transform.resize
```
Comment (lemairecarl) [inline monai/transforms/spatial/array.py:687]: typo: downsampling

File: monai/transforms/spatial/array.py (hunk 1 above)
Comment (Can-Zhao) [reply monai/transforms/spatial/array.py:601]: Oh, I didn't saw them. I will reuse one of them. Thank you!!

File: monai/transforms/spatial/array.py (hunk 2 above)
Comment (Can-Zhao) [reply monai/transforms/spatial/array.py]: Thanks!

**Comment by wyli** [comment]
thanks, it looks good, could you help add a test case here https://github.com/Project-MONAI/MONAI/blob/1516ca758090dd8ed2890b64554a77f681a0e224/tests/test_resize.py#L49-L50

ideally the output should be consistent with `anti_aliasing=True` of `skimage.transform.resize`, but I suspect there would be some implementation discrepancies. please help confirm the details and add an option for `anti_aliasing_sigma` if possible...

**Comment by Can-Zhao** [comment]
I have

**Review by wyli — APPROVED** [review]
Thanks, it looks good to me.

**Comment by wyli** [comment]
there's an issue with the tests (pytorch 1.8.1 + cuda v11.2) but I couldn't replicate it locally.
```
[2022-05-03T22:28:08.278Z] ======================================================================
[2022-05-03T22:28:08.278Z] FAIL: test_correct_results_2 (tests.test_resize.TestResize)
[2022-05-03T22:28:08.278Z] ----------------------------------------------------------------------
[2022-05-03T22:28:08.278Z] Traceback (most recent call last):
[2022-05-03T22:28:08.278Z]   File "/opt/conda/lib/python3.8/site-packages/parameterized/parameterized.py", line 533, in standalone_func
[2022-05-03T22:28:08.278Z]     return func(*(a + p.args), **p.kwargs)
[2022-05-03T22:28:08.278Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_resize.py", line 67, in test_correct_results
[2022-05-03T22:28:08.278Z]     assert_allclose(out, expected, type_test=False, atol=0.9)
[2022-05-03T22:28:08.278Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/utils.py", line 107, in assert_allclose
[2022-05-03T22:28:08.278Z]     np.testing.assert_allclose(actual, desired, *args, **kwargs)
[2022-05-03T22:28:08.278Z]   File "/opt/conda/lib/python3.8/site-packages/numpy/testing/_private/utils.py", line 1527, in assert_allclose
[2022-05-03T22:28:08.278Z]     assert_array_compare(compare, actual, desired, err_msg=str(err_msg),
[2022-05-03T22:28:08.278Z]   File "/opt/conda/lib/python3.8/site-packages/numpy/testing/_private/utils.py", line 840, in assert_array_compare
[2022-05-03T22:28:08.278Z]     raise AssertionError(msg)
[2022-05-03T22:28:08.278Z] AssertionError: 
[2022-05-03T22:28:08.278Z] Not equal to tolerance rtol=1e-07, atol=0.9
[2022-05-03T22:28:08.278Z] 
[2022-05-03T22:28:08.278Z] Mismatched elements: 2528 / 32768 (7.71%)
[2022-05-03T22:28:08.278Z] Max absolute difference: 1.
[2022-05-03T22:28:08.278Z] Max relative difference: 1.
[2022-05-03T22:28:08.278Z]  x: array([[[[0., 0., 0., ..., 0., 0., 0.],
[2022-05-03T22:28:08.278Z]          [0., 0., 0., ..., 0., 0., 0.],
[2022-05-03T22:28:08.278Z]          [0., 0., 0., ..., 0., 0., 0.],...
[2022-05-03T22:28:08.278Z]  y: array([[[[0., 0., 0., ..., 0., 0., 0.],
[2022-05-03T22:28:08.278Z]          [0., 0., 0., ..., 0., 0., 0.],
[2022-05-03T22:28:08.278Z]          [0., 0., 0., ..., 0., 0., 0.],...
[2022-05-03T22:28:08.278Z] 
[2022-05-03T22:28:08.278Z] ----------------------------------------------------------------------
```

**Comment by wyli** [comment]
there's an issue with the tests (pytorch 1.8.1 + cuda v11.2) but I couldn't replicate it locally.
```
[2022-05-03T22:28:08.278Z] ======================================================================
[2022-05-03T22:28:08.278Z] FAIL: test_correct_results_2 (tests.test_resize.TestResize)
[2022-05-03T22:28:08.278Z] ----------------------------------------------------------------------
[2022-05-03T22:28:08.278Z] Traceback (most recent call last):
[2022-05-03T22:28:08.278Z]   File "/opt/conda/lib/python3.8/site-packages/parameterized/parameterized.py", line 533, in standalone_func
[2022-05-03T22:28:08.278Z]     return func(*(a + p.args), **p.kwargs)
[2022-05-03T22:28:08.278Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_resize.py", line 67, in test_correct_results
[2022-05-03T22:28:08.278Z]     assert_allclose(out, expected, type_test=False, atol=0.9)
[2022-05-03T22:28:08.278Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/utils.py", line 107, in assert_allclose
[2022-05-03T22:28:08.278Z]     np.testing.assert_allclose(actual, desired, *args, **kwargs)
[2022-05-03T22:28:08.278Z]   File "/opt/conda/lib/python3.8/site-packages/numpy/testing/_private/utils.py", line 1527, in assert_allclose
[2022-05-03T22:28:08.278Z]     assert_array_compare(compare, actual, desired, err_msg=str(err_msg),
[2022-05-03T22:28:08.278Z]   File "/opt/conda/lib/python3.8/site-packages/numpy/testing/_private/utils.py", line 840, in assert_array_compare
[2022-05-03T22:28:08.278Z]     raise AssertionError(msg)
[2022-05-03T22:28:08.278Z] AssertionError: 
[2022-05-03T22:28:08.278Z] Not equal to tolerance rtol=1e-07, atol=0.9
[2022-05-03T22:28:08.278Z] 
[2022-05-03T22:28:08.278Z] Mismatched elements: 2528 / 32768 (7.71%)
[2022-05-03T22:28:08.278Z] Max absolute difference: 1.
[2022-05-03T22:28:08.278Z] Max relative difference: 1.
[2022-05-03T22:28:08.278Z]  x: array([[[[0., 0., 0., ..., 0., 0., 0.],
[2022-05-03T22:28:08.278Z]          [0., 0., 0., ..., 0., 0., 0.],
[2022-05-03T22:28:08.278Z]          [0., 0., 0., ..., 0., 0., 0.],...
[2022-05-03T22:28:08.278Z]  y: array([[[[0., 0., 0., ..., 0., 0., 0.],
[2022-05-03T22:28:08.278Z]          [0., 0., 0., ..., 0., 0., 0.],
[2022-05-03T22:28:08.278Z]          [0., 0., 0., ..., 0., 0., 0.],...
[2022-05-03T22:28:08.278Z] 
[2022-05-03T22:28:08.278Z] ----------------------------------------------------------------------
```

**Comment by Can-Zhao** [comment]
> there's an issue with the tests (pytorch 1.8.1 + cuda v11.2) but I couldn't replicate it locally.
> 
> ```
> [2022-05-03T22:28:08.278Z] ======================================================================
> [2022-05-03T22:28:08.278Z] FAIL: test_correct_results_2 (tests.test_resize.TestResize)
> [2022-05-03T22:28:08.278Z] ----------------------------------------------------------------------
> [2022-05-03T22:28:08.278Z] Traceback (most recent call last):
> [2022-05-03T22:28:08.278Z]   File "/opt/conda/lib/python3.8/site-packages/parameterized/parameterized.py", line 533, in standalone_func
> [2022-05-03T22:28:08.278Z]     return func(*(a + p.args), **p.kwargs)
> [2022-05-03T22:28:08.278Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_resize.py", line 67, in test_correct_results
> [2022-05-03T22:28:08.278Z]     assert_allclose(out, expected, type_test=False, atol=0.9)
> [2022-05-03T22:28:08.278Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/utils.py", line 107, in assert_allclose
> [2022-05-03T22:28:08.278Z]     np.testing.assert_allclose(actual, desired, *args, **kwargs)
> [2022-05-03T22:28:08.278Z]   File "/opt/conda/lib/python3.8/site-packages/numpy/testing/_private/utils.py", line 1527, in assert_allclose
> [2022-05-03T22:28:08.278Z]     assert_array_compare(compare, actual, desired, err_msg=str(err_msg),
> [2022-05-03T22:28:08.278Z]   File "/opt/conda/lib/python3.8/site-packages/numpy/testing/_private/utils.py", line 840, in assert_array_compare
> [2022-05-03T22:28:08.278Z]     raise AssertionError(msg)
> [2022-05-03T22:28:08.278Z] AssertionError: 
> [2022-05-03T22:28:08.278Z] Not equal to tolerance rtol=1e-07, atol=0.9
> [2022-05-03T22:28:08.278Z] 
> [2022-05-03T22:28:08.278Z] Mismatched elements: 2528 / 32768 (7.71%)
> [2022-05-03T22:28:08.278Z] Max absolute difference: 1.
> [2022-05-03T22:28:08.278Z] Max relative difference: 1.
> [2022-05-03T22:28:08.278Z]  x: array([[[[0., 0., 0., ..., 0., 0., 0.],
> [2022-05-03T22:28:08.278Z]          [0., 0., 0., ..., 0., 0., 0.],
> [2022-05-03T22:28:08.278Z]          [0., 0., 0., ..., 0., 0., 0.],...
> [2022-05-03T22:28:08.278Z]  y: array([[[[0., 0., 0., ..., 0., 0., 0.],
> [2022-05-03T22:28:08.278Z]          [0., 0., 0., ..., 0., 0., 0.],
> [2022-05-03T22:28:08.278Z]          [0., 0., 0., ..., 0., 0., 0.],...
> [2022-05-03T22:28:08.278Z] 
> [2022-05-03T22:28:08.278Z] ----------------------------------------------------------------------
> ```
I was wondering if you could reproduce this error. I could not... I was suspecting that some randomly generated test data gives different result for skimage and our implementation.

**Pull request closed without merging**